### ISSUE TITLE
mark/reset not supported / Stream closed

### ISSUE THREAD

**Issue by mtsr** [junkdog/artemis-odb#315]
Passing a FileInputStream to WorldSerializationManager#load gives an IOException:

     java.lang.RuntimeException: java.io.IOException: mark/reset not supported
        at com.artemis.io.JsonArtemisSerializer.updateLookupMap(JsonArtemisSerializer.java:86)
        at com.artemis.io.JsonArtemisSerializer.load(JsonArtemisSerializer.java:71)
        at com.artemis.managers.WorldSerializationManager.load(WorldSerializationManager.java:30)
        at nl.jonasmatser.solar.screens.GameScreen.<init>(GameScreen.java:66)
        at nl.jonasmatser.solar.Solar.create(Solar.java:9)
        at com.badlogic.gdx.backends.lwjgl.LwjglApplication.mainLoop(LwjglApplication.java:143)
        at com.badlogic.gdx.backends.lwjgl.LwjglApplication$1.run(LwjglApplication.java:120)
    Caused by: java.io.IOException: mark/reset not supported
        at java.io.InputStream.reset(InputStream.java:347)
        at com.artemis.io.JsonArtemisSerializer.updateLookupMap(JsonArtemisSerializer.java:83)
        ... 6 more

Decorating it with a BufferedInputStream gives another IOException:

    java.lang.RuntimeException: java.io.IOException: Stream closed
        at com.artemis.io.JsonArtemisSerializer.updateLookupMap(JsonArtemisSerializer.java:86)
        at com.artemis.io.JsonArtemisSerializer.load(JsonArtemisSerializer.java:71)
        at com.artemis.managers.WorldSerializationManager.load(WorldSerializationManager.java:30)
        at nl.jonasmatser.solar.screens.GameScreen.<init>(GameScreen.java:67)
        at nl.jonasmatser.solar.Solar.create(Solar.java:9)
        at com.badlogic.gdx.backends.lwjgl.LwjglApplication.mainLoop(LwjglApplication.java:143)
        at com.badlogic.gdx.backends.lwjgl.LwjglApplication$1.run(LwjglApplication.java:120)
    Caused by: java.io.IOException: Stream closed
        at java.io.BufferedInputStream.getBufIfOpen(BufferedInputStream.java:162)
        at java.io.BufferedInputStream.reset(BufferedInputStream.java:435)
        at com.artemis.io.JsonArtemisSerializer.updateLookupMap(JsonArtemisSerializer.java:83)
        ... 6 more

The only reason the tests currently work is that .close() doesn't have an effect on a ByteArrayInputStream.

**Comment by mtsr** [comment]
Also note that something like a BufferedInputStream can still throw an exception on mark+reset if the number of bytes since mark is bigger than the buffer.

**Comment by junkdog** [comment]
Meh... why isn't it supported... silly. I assume the error arises because we have to read the componentIdentifiers prior to deserializing the entities, effectively reading the stream twice.

mark() isn't used.

**Comment by mtsr** [comment]
^^ it isn't supported because the data can very well not be in memory any more, as is the case with a BufferedInputStream that is too small. As to why FileInputStreams don't support mark/reset? Good question.

**Comment by mtsr** [comment]
Unfortunately that means either being smarter about reading the data only once, or copying the whole thing into a ByteArray, which could be prohibitive in the case of large worlds.

**Comment by mtsr** [comment]
[This answer](http://stackoverflow.com/a/30437553/1152077) has some alternatives.

**Comment by junkdog** [comment]
Yeah, but it'd be nice if we there was a way for FileInputStream to re-open the stream.

I don't think jsonbeans can be coaxed into reading the elements in a predefined order, without pr...

_new comment is posted_

oh - thanks! Yeah, that's def the path of least resistance!

**Comment by junkdog** [comment]
not the nicest solution, but path of least resistance.

**Comment by mtsr** [comment]
Unfortunately there's another issue in there. Close is called on the inputstream during the first read as well. This is OK with a byte array input stream, as it doesn't actually do anything, but with a proper mark supporting stream, this still means an error when using the stream after that. Now on my phone, but will look up the line number later.

**Comment by junkdog** [comment]
Hmm, I briefly looked at the code - I think JsonBeans itself is closing the stream (I could be wrong; still on my ~~first~~ second cup of :coffee:). I'll see if there's a way around it.
